### PR TITLE
i#3101 drcachesim: add binary interface for running separately-built analysis tools

### PR DESCRIPTION
**Pull request by yuriershov**
This pull request includes the following functionality:
 
 - Added support of loading separately-built analysis tools to drcachesim dynamically
 - Added opportunity to run multiple simulators simultaneously

Issue: https://github.com/DynamoRIO/dynamorio/issues/3101

### CODE REVIEW

**Comment by derekbruening** [comment]
Looks like there are still some build issues breaking some platforms.  Please indicate when this is ready for review.

File: clients/drcachesim/tools/loader/external_tool_creator.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+class external_tool_creator : public dynamic_lib {
+public:
+    using get_tool_id_t = const char *(*)();
+    using create_tool_t = analysis_tool_t *(*)();
+
+    const get_tool_id_t get_id_;
+    const create_tool_t create_tool_;
+
+    external_tool_creator(const char *filename);
```
Comment (derekbruening) [inline clients/drcachesim/tools/loader/external_tool_creator.h:50]: Generally we have the consructor first.

Comment (derekbruening) [inline clients/drcachesim/tools/loader/external_tool_creator.h:50]: Why not take in `const std::string &`?

File: clients/drcachesim/tools/loader/external_tool_creator.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+namespace dynamorio {
+namespace drmemtrace {
+
+external_tool_creator::external_tool_creator(const char *filename)
+    : dynamic_lib(filename)
+    , get_id_(load<get_tool_id_t>("get_id"))
+    , create_tool_(load<create_tool_t>("analysis_tool_create"))
+{
+    if (create_tool_ == NULL)
+        throw dynamic_lib_error(error());
```
Comment (derekbruening) [inline clients/drcachesim/tools/loader/external_tool_creator.cpp:44]: As above, no C++ exceptions.

Comment (derekbruening) [inline clients/drcachesim/tools/loader/external_tool_creator.cpp:44]: What if get_id_ or create_tool_ are nullptr?

File: clients/drcachesim/analyzer_multi.cpp [hunk 3]
```diff
@@ -60,13 +62,50 @@
...
+    for (; iter != end; ++iter) {
+        if ((*iter).find(".drcachesim") != std::string::npos) {
+            std::string abs_path(tools_dir);
+            abs_path.append(*iter);
+            external_tool_config_file_t config(op_dr_root.get_value(), abs_path);
+            if (config.valid_ && config.id_ == id) {
+                try {
+                    external_tool_creator creator(config.creator_path_.c_str());
+                    DR_ASSERT(creator.get_id_() != id.c_str());
+                    tool = creator.create_tool_();
```
Comment (derekbruening) [inline clients/drcachesim/analyzer_multi.cpp:97]: See comment below: this is currently not guaranteed to not be nullptr; if not checked elsewhere, should check here.

File: clients/drcachesim/tools/external/example/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+TOOL_ID=${id}
+CREATOR_BIN=${libdir}/${LIB_PFX}${id}_creator${LIB_EXT}\n\
+")
+endfunction ()
+
+set(EMPTY_CONFIG_INSTALL ${PROJECT_BINARY_DIR}/empty.drcachesim)
+set(EMPTY_CONFIG_BUILD ${PROJECT_BINARY_DIR}/tools/empty.drcachesim)
+write_config_file(empty ${EMPTY_CONFIG_BUILD} clients/${INSTALL_LIB})
+write_config_file(empty ${EMPTY_CONFIG_INSTALL} tools/${INSTALL_LIB})
+DR_install(FILES "${EMPTY_CONFIG_INSTALL}" DESTINATION ${INSTALL_CLIENTS_BASE})
```
Comment (derekbruening) [inline clients/drcachesim/tools/external/example/CMakeLists.txt:117]: Please add a test of running this sample external tool.

**Review by derekbruening — COMMENTED** [review]
Thank you for the contribution.  This is a nice feature.  I mostly have a bunch of minor style comments; nothing major beyond adding more docs and adding a test.

File: clients/drcachesim/analyzer_multi.cpp [hunk 5]
```diff
@@ -203,6 +242,7 @@ analyzer_multi_t::~analyzer_multi_t()
     }
 #endif
     destroy_analysis_tools();
+    loaders_.clear();
```
[the remark this answers is not in the document]
Comment (yuriershov) [reply clients/drcachesim/analyzer_multi.cpp]: Yes, you're right. I've removed this.

File: clients/drcachesim/tools/external/example/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+# We need to pass /EHsc and we pull in libcmtd into drcachesim from a dep lib.
+# Thus we need to override the /MT with /MTd.
+#
+# TODO i#6040: We should add a new configure_original_flage(target) function
+# which restores the ORIG_ flags at the source level for all other targets
+# in this file which are neither clients nor use DR standalone. We can then
+# get rid of add_win32_flags(target) since that just adds a couple of flags
+# back, and using the ORIG_ values is preferable.
+macro(add_win32_flags target)
```
[the remark this answers is not in the document]
Comment (yuriershov) [reply clients/drcachesim/tools/external/example/CMakeLists.txt]: I assumed there should be some full-package analysis tool as a reference, which should contain all the scope (creator, <concrete>_tool_create factory function, launcher). So user can get it, made necessary changes and build without and independant on drcachesim cmake configs.
I removed code duplicates, but probably it's make sense to put these macros to some separate .cmake file to share this functionality.

File: clients/drcachesim/tools/external/example/empty_launcher.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
```
[the remark this answers is not in the document]
Comment (yuriershov) [reply clients/drcachesim/tools/external/example/empty_launcher.cpp:32]: I mentioned earlier the purpose of this launcher.
The test has been added.

File: clients/drcachesim/analyzer_multi.cpp (hunk 3 above)
Comment (yuriershov) [reply clients/drcachesim/analyzer_multi.cpp]: There is checking for nullptr after this function.

File: clients/drcachesim/tools/external/example/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/external/example/CMakeLists.txt]: These Windows flags changes are not necessary for a standalone cmake setup: they are only needed when embedded inside the DR cmake which completely redoes all the build flags.  If removing this completely from external/ has no build issues let's go with that as we do not expect to need debug build here.

File: clients/drcachesim/analyzer_multi.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/analyzer_multi.cpp]: OK.

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 8]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
 tlb_simulator_create(), func_view_create(), and syscall_mix_tool_create()
 functions.
 
+\section external_tools Separately-built tools
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1606]: nit: The other sections capitalize each word: so "Separately-Built Tools"

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 9]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+non-predefined separately-built external tools. This tool can be loaded by drcachesim
+using the \p -simulator_type option.
+
+## External tool package
+
+The tool package should consist of
+- <tt>Registration file</tt> called \p toolname.drcachesim.
+- <tt>Static library</tt> containing subclass
+#dynamorio::drmemtrace::analysis_tool_t with tool internal logic. This library described
+in prevoius section.
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1618]: spelling: previous

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 10]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+DynamoRIO installation.  Here \p toolname is the desired external name of the tool.
+This file should contain the following lines:
+
+\code
+TOOL_NAME=name-of-tool
+CREATOR_BIN32=/absolute/path/to/32-bit-creator-library
+CREATOR_BIN64=/absolute/path/to/64-bit-creator-library
+\endcode
+
+This enables \p drcachsim to locate the tool's creator library.  The 32 and
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1631]: spelling: drcachesim

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 11]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+
+\p drcachesim \p drmemtrace analysis tool framework allows to load
+non-predefined separately-built external tools. This tool can be loaded by drcachesim
+using the \p -simulator_type option.
+
+## External tool package
+
+The tool package should consist of
+- <tt>Registration file</tt> called \p toolname.drcachesim.
+- <tt>Static library</tt> containing subclass
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1616]: nit: "a subclass of"

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 12]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+#dynamorio::drmemtrace::analysis_tool_t with tool internal logic. This library described
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1617]: nit: "library was"

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 13]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+
+This enables \p drcachsim to locate the tool's creator library.  The 32 and
+64 specifiers allow pointing at alternate-bitwidth paths for use if
+the target application creates a child process of a different bitwidth.
+
+For more extensive actions on launching the tool, a custom front-end
+executable can be created that replaces \p drcachesim modeled after histogram_launcher.cpp
+or opcode_mix_launcher.cpp.
+
+Creator dynamic library should contain 2 export functions:
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1639]: nit: "The creator"

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 14]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+
+\code
+extern "C" EXPORT const char *
+get_tool_name()
+{
+    return "name-of-concrete-tool";
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1645]: nit: Above it was "name-of-tool": best to use the same name else this looks like the user picks two different names, a regular and a "concrete" name.

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 15]
```diff
@@ -1603,6 +1603,60 @@ reuse_time_tool_create(), view_tool_create(), cache_simulator_create(),
...
+}
+
+extern "C" EXPORT analysis_tool_t *
+analysis_tool_create()
+{
+    return <concrete>_tool_create();
+}
+\endcode
+
+which allows \p drcachsim to create appropriate analyis tool.  As an
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1655]: spelling: "drcachesim"

Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:1655]: nit: "an analysis tool"

File: clients/drcachesim/tools/external/example/CMakeLists.txt [hunk 16]
```diff
@@ -115,3 +82,36 @@ set(EMPTY_CONFIG_BUILD ${PROJECT_BINARY_DIR}/tools/empty.drcachesim)
...
+    PROPERTIES PASS_REGULAR_EXPRESSION "${empty_regex}")
+
+  # Simple test to ensure non-existent tool load by drcachesim should fail.
+  file(READ "${CMAKE_CURRENT_SOURCE_DIR}/nonexistent.templatex" nonexistent_regex)
+  add_test(NAME drcachesim.non-existent_load
+    COMMAND ${PROJECT_BINARY_DIR}/bin64/drrun -t drcachesim -offline
+     -simulator_type non-existent -indir ${trace_dir})
+  set_tests_properties(drcachesim.non-existent_load
+    PROPERTIES PASS_REGULAR_EXPRESSION "${nonexistent_regex}")
+endif ()
```
Comment (derekbruening) [inline clients/drcachesim/tools/external/example/CMakeLists.txt:117]: nit: Missing newline

File: clients/drcachesim/tools/external/example/CMakeLists.txt [hunk 17]
```diff
@@ -115,3 +82,36 @@ set(EMPTY_CONFIG_BUILD ${PROJECT_BINARY_DIR}/tools/empty.drcachesim)
...
+if (X86 AND X64 AND ZIP_FOUND)
+# Simple test to ensure empty launcher keeps working.
+  set(trace_dir
+    "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/drmemtrace.threadsig.x64.tracedir")
+  add_test(NAME tool.empty_launcher
+    COMMAND empty_launcher -trace ${trace_dir})
+  set_tests_properties(tool.empty_launcher PROPERTIES TIMEOUT ${test_seconds})
+
+  # Simple test to ensure empty tool can be loaded by drcachesim.
+  file(READ "${CMAKE_CURRENT_SOURCE_DIR}/empty.templatex" empty_regex)
```
Comment (derekbruening) [inline clients/drcachesim/tools/external/example/CMakeLists.txt:101]: For these simple matches meant to work as an external project and so not using the DR test .templatex infrastructure, we could simplify further by removing the file and just putting the two line string directly here as a literal.  Ditto for the other file.

File: clients/drcachesim/tools/external/example/empty.cpp [hunk 18]
```diff
@@ -111,5 +110,6 @@ empty_t::process_memref(const memref_t &memref)
 bool
 empty_t::print_results()
 {
+    std::cout << "Empty tool results:" << std::endl;
```
Comment (derekbruening) [inline clients/drcachesim/tools/external/example/empty.cpp:113]: Tools print to stderr by convention (for online use where output is mixed with app output, to distinguish from the app).

File: clients/drcachesim/tools/external/example/empty.h [hunk 19]
```diff
@@ -40,8 +42,7 @@ using dynamorio::drmemtrace::memref_t;
 
 class empty_t : public analysis_tool_t {
 public:
-    empty_t(const std::string &module_file_path, unsigned int verbose,
-            const std::string &alt_module_dir = "");
+    empty_t(unsigned int verbose);
```
Comment (derekbruening) [inline clients/drcachesim/tools/external/example/empty.h:45]: Add `explicit` keyword for single-arg constructor.

File: clients/drcachesim/tools/loader/external_tool_creator.h [hunk 20]
```diff
@@ -39,15 +43,20 @@
...
+class external_tool_creator_t : public dynamic_lib_t {
 public:
-    using get_tool_id_t = const char *(*)();
+    external_tool_creator_t(const std::string &filename);
+    std::string
+    get_tool_name();
+    analysis_tool_t *
+    create_tool();
+
+public:
```
Comment (derekbruening) [inline clients/drcachesim/tools/loader/external_tool_creator.h:54]: If these fields now have accessors (right above) they can private?

**Review by derekbruening — APPROVED** [review]
Looks good, just some spelling/writing nits.

**Comment by derekbruening** [comment]
In the future, please reply to each comment with "Done" or something similar, rather than just marking resolved with no reply, as documented: https://dynamorio.org/page_code_reviews.html#autotoc_md114

**Pull request merged**